src/pal/src/sharedmemory/sharedmemory.cpp

```diff
@@ -0,0 +1,1114 @@
+// SharedMemoryHelpers
+
+const mode_t SharedMemoryHelpers::PermissionsMask_AllUsers_ReadWrite =
+    S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH;
+const mode_t SharedMemoryHelpers::PermissionsMask_AllUsers_ReadWriteExecute =
+    PermissionsMask_AllUsers_ReadWrite | (S_IXUSR | S_IXGRP | S_IXOTH);
+const UINT32 SharedMemoryHelpers::InvalidPid = static_cast<UINT32>(-1);
+
+void *SharedMemoryHelpers::Alloc(SIZE_T byteCount)
+{
+    void *buffer = InternalMalloc(byteCount);
+    if (buffer == nullptr)
+    {
+        throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::OutOfMemory));
+    }
+    return buffer;
+}
+
+SIZE_T SharedMemoryHelpers::AlignDown(SIZE_T value, SIZE_T alignment)
+{
+    _ASSERTE((alignment & (alignment - 1)) == 0); // must be a power of 2
+    return value & ~(alignment - 1);
+}
+
+SIZE_T SharedMemoryHelpers::AlignUp(SIZE_T value, SIZE_T alignment)
+{
+    _ASSERTE((alignment & (alignment - 1)) == 0); // must be a power of 2
+    return AlignDown(value + (alignment - 1), alignment);
+}
+
+bool SharedMemoryHelpers::EnsureDirectoryExists(const char *path, bool createIfNotExist)
+{
+    // Check if the path already exists
+    struct stat statInfo;
+    int statResult = stat(path, &statInfo);
+    if (statResult != 0 && errno == ENOENT)
+    {
+        if (!createIfNotExist)
+        {
+            return false;
+        }
+
+        // The path does not exist, create the directory
+        if (mkdir(path, PermissionsMask_AllUsers_ReadWriteExecute) == 0)
+        {
+            // The permissions mask passed to mkdir() is filtered by the process' permissions umask, so mkdir() may not set all
+            // of the requested permissions. Use chmod() to set the proper permissions.
+            if (chmod(path, PermissionsMask_AllUsers_ReadWriteExecute) == 0)
+            {
+                return true;
+            }
+            rmdir(path);
+            throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::IO));
+        }
+        if (errno != EEXIST)
+        {
+            throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::IO));
+        }
+
+        statResult = stat(path, &statInfo);
+    }
+
+    // The path exists, check that it's a directory
+    if (statResult != 0 || !(statInfo.st_mode & S_IFDIR))
+    {
+        throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::IO));
+    }
+
+    // Check the directory's permissions and try to update them
+    if ((statInfo.st_mode & PermissionsMask_AllUsers_ReadWriteExecute) == PermissionsMask_AllUsers_ReadWriteExecute)
+    {
+        return true;
+    }
+    if (!createIfNotExist || chmod(path, PermissionsMask_AllUsers_ReadWriteExecute) != 0)
+    {
+        throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::IO));
+    }
+    return true;
+}
+
+int SharedMemoryHelpers::OpenDirectory(LPCSTR path)
```

Comment:
The OpenDirectory and CreateOrOpenFile have almost the same block of code for handling errors repeated three times. I would recommend refactoring that into a helper "Open" function that would handle the EINTR internally, contain the throws and only return if the status was ok or ENOENT.